src/rnnlm/rnnlm-simple-looped.h

```diff
@@ -0,0 +1,184 @@
+    optimize_config.Register(&optimization_opts);
+
+    // register the compute options with the prefix "computation".
+    ParseOptions compute_opts("computation", opts);
+    compute_config.Register(&compute_opts);
+  }
+};
+
+class RnnlmSimpleLoopedInfo  {
+ public:
+  RnnlmSimpleLoopedInfo(
+      const RnnlmSimpleLoopedComputationOptions &opts,
+      const kaldi::nnet3::Nnet &rnnlm,
+      const CuMatrix<BaseFloat> &word_embedding_mat);
+
+  void Init(const RnnlmSimpleLoopedComputationOptions &opts,
+            const kaldi::nnet3::Nnet &rnnlm,
+            const CuMatrix<BaseFloat> &word_embedding_mat);
+
+  const RnnlmSimpleLoopedComputationOptions &opts;
+
+  const kaldi::nnet3::Nnet &rnnlm;
+  const CuMatrix<BaseFloat> &word_embedding_mat;
+
+  // frames_left_context equals the model left context plus the value of the
+  // --extra-left-context-initial option.
+  int32 frames_left_context;
+  // frames_right_context is the same as the right-context of the model.
+  int32 frames_right_context;
+  // The frames_per_chunk equals the number of input frames we need for each
+  // chunk (except for the first chunk).
+  int32 frames_per_chunk;
+
+  // The output dimension of the nnet neural network (not the final output).
+  int32 nnet_output_dim;
+
+  // The 3 computation requests that are used to create the looped
+  // computation are stored in the class, as we need them to work out
+  // exactly shich iVectors are needed.
+  ComputationRequest request1, request2, request3;
+
+  // The compiled, 'looped' computation.
+  NnetComputation computation;
+};
+
+/*
+  This class handles the neural net computation; it's mostly accessed
+  via other wrapper classes.
+
+  It accept just input features */
+class RnnlmSimpleLooped {
+ public:
+  /**
+     This constructor takes features as input.
+     Note: it stores references to all arguments to the constructor, so don't
+     delete them till this goes out of scope.
+
+     @param [in] info   This helper class contains all the static pre-computed information
+                        this class needs, and contains a pointer to the neural net.
+     @param [in] feats  The input feature matrix.
+  */
+  RnnlmSimpleLooped(const RnnlmSimpleLoopedInfo &info);
+
+  RnnlmSimpleLooped(const RnnlmSimpleLooped &other);
+
+  inline int32 NnetOutputDim() const { return info_.nnet_output_dim; }
+
+  // Gets the nnet's output for a particular frame, with 0 <= frame < NumFrames().
+  // 'output' must be correctly sized (with dimension NnetOutputDim()).  Note:
+  // you're expected to call this, and GetOutput(), in an order of increasing
+  // frames.  If you deviate from this, one of these calls may crash.
+//  void GetNnetOutputForFrame(int32 frame, VectorBase<BaseFloat> *output);
+
+  // Updates feats_ with the new incoming word specified in word_indexes
+  // We usually do this one at a time
+  void TakeFeatures(const std::vector<int32> &word_indexes);
```

Comment:
You always advance this one word at a time, so I think it would make more sense to have an interface like
```
/// Advance the state of the RNNLM by appending this word to the word sequence.
void AddWord(int32 word);

/// Return the log-prob that the model predicts for the provided word-index,
/// given the previous history determined by the sequence of calls to AddWord()
/// (implicitly starting with the BOS symbol).
BaseFloat LogProbOfWord(int32 word_index) const;
```

The class itself can internally maintain a pointer (maybe just a BaseFloat* pointing to GPU memory) 
to the vector consisting of the neural network's output at the current position.   This can be a pointer
into the matrix of output obtained by calling GetOutput() with the non-destructive version.
There is no need to expose the vector that the neural network outputs.